File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/Commands/Command/Start.java
package jp.aoichaan0513.A_TosoGame_Live.Commands.Command;

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.GameManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI;
import jp.aoichaan0513.A_TosoGame_Live.Commands.ICommand;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import jp.aoichaan0513.A_TosoGame_Live.Timer.Timer;
import org.bukkit.Bukkit;
import org.bukkit.command.BlockCommandSender;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.entity.Player;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class Start extends ICommand {

    public Start(String name) {
        super(name);
    }

    @Override
    public void onPlayerCommand(Player sp, Command cmd, String label, String[] args) {
        if (TosoGameAPI.isAdmin(sp)) {
            if (!GameManager.isGame()) {
                WorldConfig worldConfig = Main.getWorldConfig();

                if (args.length != 0) {
                    if (args[0].equalsIgnoreCase("skip")) {
                        Timer.start(1, worldConfig.getGameConfig().getGame());
                        return;
                    } else {
                        try {
                            int i = Integer.parseInt(args[0]);
                            Bukkit.broadcastMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "カウントダウン開始");
                            Timer.start(i + 1, worldConfig.getGameConfig().getGame());
                            return;
                        } catch (NumberFormatException ex) {
                            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "数字を指定してください。");
                            return;
                        }
                    }
                }
                Bukkit.broadcastMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "カウントダウン開始");
                Timer.start(worldConfig.getGameConfig().getCountDown(), worldConfig.getGameConfig().getGame());
                return;
            }
            sp.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "ゲームが実行されているため実行できません。");
            return;
        }
        MainAPI.sendMessage(sp, MainAPI.ErrorMessage.PERMISSIONS);
        return;
    }

    @Override
    public void onBlockCommand(BlockCommandSender bs, Command cmd, String label, String[] args) {
        if (!GameManager.isGame()) {
            WorldConfig worldConfig = Main.getWorldConfig();

            if (args.length != 0) {
                if (args[0].equalsIgnoreCase("skip")) {
                    Timer.start(1, worldConfig.getGameConfig().getGame());
                    return;
                } else {
                    try {
                        int i = Integer.parseInt(args[0]);
                        Bukkit.broadcastMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "カウントダウン開始");
                        Timer.start(i + 1, worldConfig.getGameConfig().getGame());
                        return;
                    } catch (NumberFormatException ex) {
                        bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "数字を指定してください。");
                        return;
                    }
                }
            }
            Bukkit.broadcastMessage(MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "カウントダウン開始");
            Timer.start(worldConfig.getGameConfig().getCountDown(), worldConfig.getGameConfig().getGame());
            return;
        }
        bs.sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "ゲームが実行されているため実行できません。");
        return;
    }

    @Override
    public void onConsoleCommand(ConsoleCommandSender cs, Command cmd, String label, String[] args) {
        MainAPI.sendMessage(cs, MainAPI.ErrorMessage.NOT_PLAYER);
        return;
    }

    @Override
    public List<String> onPlayerTabComplete(Player sp, Command cmd, String alias, String[] args) {
        return null;
    }

    @Override
    public List<String> onBlockTabComplete(BlockCommandSender bs, Command cmd, String alias, String[] args) {
        return null;
    }

    @Override
    public List<String> onConsoleTabComplete(ConsoleCommandSender cs, Command cmd, String alias, String[] args) {
        return null;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
        if (!command.getName().equalsIgnoreCase("start")) return null;
        if (args.length == 1) {
            if (args[0].length() == 0) {
                return Arrays.asList("skip");
            } else {
                if ("skip".startsWith(args[0])) {
                    return Collections.singletonList("skip");
                }
            }
        }
        return null;
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/Commands/Command/Start.kt
package jp.aoichaan0513.A_TosoGame_Live.Commands.Command

import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.ErrorMessage
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.GameManager
import jp.aoichaan0513.A_TosoGame_Live.API.TosoGameAPI
import jp.aoichaan0513.A_TosoGame_Live.Commands.ICommand
import jp.aoichaan0513.A_TosoGame_Live.Main
import jp.aoichaan0513.A_TosoGame_Live.Utils.ParseUtil
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.BlockCommandSender
import org.bukkit.command.Command
import org.bukkit.command.ConsoleCommandSender
import org.bukkit.entity.Player

class Start(name: String) : ICommand(name) {

    override fun onPlayerCommand(sp: Player, cmd: Command, label: String, args: Array<String>) {
        if (TosoGameAPI.isAdmin(sp)) {
            if (!GameManager.isGame()) {
                if (GameManager.isGame(GameManager.GameState.NONE)) {
                    val worldConfig = Main.worldConfig
                    if (args.isNotEmpty()) {
                        if (args[0].equals("skip", true)) {
                            GameManager.startGame(1, worldConfig.gameConfig.game)
                            return
                        } else {
                            if (ParseUtil.isInt(args[0])) {
                                val i = args[0].toInt()
                                Bukkit.broadcastMessage("${MainAPI.getPrefix(PrefixType.SECONDARY)}カウントダウン開始")
                                GameManager.startGame(i + 1, worldConfig.gameConfig.game)
                                return
                            } else {
                                MainAPI.sendMessage(sp, ErrorMessage.ARGS_INTEGER)
                                return
                            }
                        }
                    } else {
                        Bukkit.broadcastMessage("${MainAPI.getPrefix(PrefixType.SECONDARY)}カウントダウン開始")
                        GameManager.startGame(worldConfig.gameConfig.countDown, worldConfig.gameConfig.game)
                        return
                    }
                }
                sp.sendMessage("""
                    ${MainAPI.getPrefix(PrefixType.ERROR)}ゲームが終了しているため実行できません。
                    ${MainAPI.getPrefix(PrefixType.SECONDARY)}"${ChatColor.RED}${ChatColor.UNDERLINE}/reset${ChatColor.RESET}${ChatColor.GRAY}"でゲームをリセットしてから実行してください。
                """.trimIndent())
                return
            }
            MainAPI.sendMessage(sp, ErrorMessage.GAME)
            return
        }
        MainAPI.sendMessage(sp, ErrorMessage.PERMISSIONS)
    }

    override fun onBlockCommand(bs: BlockCommandSender, cmd: Command, label: String, args: Array<String>) {
        if (!GameManager.isGame()) {
            if (GameManager.isGame(GameManager.GameState.NONE)) {
                val worldConfig = Main.worldConfig
                if (args.isNotEmpty()) {
                    if (args[0].equals("skip", true)) {
                        GameManager.startGame(1, worldConfig.gameConfig.game)
                        return
                    } else {
                        if (ParseUtil.isInt(args[0])) {
                            val i = args[0].toInt()
                            Bukkit.broadcastMessage("${MainAPI.getPrefix(PrefixType.SECONDARY)}カウントダウン開始")
                            GameManager.startGame(i + 1, worldConfig.gameConfig.game)
                            return
                        } else {
                            MainAPI.sendMessage(bs, ErrorMessage.ARGS_INTEGER)
                            return
                        }
                    }
                } else {
                    Bukkit.broadcastMessage("${MainAPI.getPrefix(PrefixType.SECONDARY)}カウントダウン開始")
                    GameManager.startGame(worldConfig.gameConfig.countDown, worldConfig.gameConfig.game)
                    return
                }
            }
            bs.sendMessage("""
                ${MainAPI.getPrefix(PrefixType.ERROR)}ゲームが終了しているため実行できません。
                ${MainAPI.getPrefix(PrefixType.SECONDARY)}"${ChatColor.RED}${ChatColor.UNDERLINE}reset${ChatColor.RESET}${ChatColor.GRAY}"でゲームをリセットしてから実行してください。
            """.trimIndent())
            return
        }
        MainAPI.sendMessage(bs, ErrorMessage.GAME)
    }

    override fun onConsoleCommand(cs: ConsoleCommandSender, cmd: Command, label: String, args: Array<String>) {
        MainAPI.sendMessage(cs, ErrorMessage.NOT_PLAYER)
        return
    }

    override fun onPlayerTabComplete(sp: Player, cmd: Command, alias: String, args: Array<String>): List<String>? {
        if (!TosoGameAPI.isAdmin(sp) || args.size != 1) return null
        return getTabList(args[0], "skip")
    }

    override fun onBlockTabComplete(bs: BlockCommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        if (args.size != 1) return null
        return getTabList(args[0], "skip")
    }

    override fun onConsoleTabComplete(cs: ConsoleCommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        return null
    }
}